src/vision/status_quo/barbara_wishes_for_easy_runtime_switch.md

```diff
@@ -0,0 +1,68 @@
+
+## 🚧 Warning: Draft status 🚧
+
+This is a draft "status quo" story submitted as part of the brainstorming period. It is derived from
+real-life experiences of actual Rust users and is meant to reflect some of the challenges that Async
+Rust programmers face today.
+
+## The story
+
+[Barbara] has been working on an async codebase for the past 5 years. It is extremely mature and
+quite large (in the millions of lines of code). The product has traditionally been run on the server
+only, but now they want to move large parts to the edge. After some investigation, they realize that
+the open-source async runtime they have been using (tokio), while great for server workloads, is
+not very performant on the low-end edge devices.
+
+Tokio makes tradeoffs in memory for performance and while performance is still a concern on
+the low-end edge devices Barbara is targeting, memory consumption is an even harder constraint there.
+Additionally, the usage of a thread pool by tokio does not help either as the targeted edge
+devices have either a single CPU thread or only a few.
+
+They had never planned to run on edge devices and so it was assumed that they would always stick
+to tokio. Barbara looks into the how much work it would be to convert their codebase from
+tokio to a custom runtime tuned to their workload. To their disappointment, they discover many
+areas, where their choice of runtime permeates the code base:
+
+* tokio provides variants of helpers macros and types, like `tokio::select!` and `tokio::Mutex`. These helpers can be used without the rest of tokio, and there are also alternatives from the futures crate and elsewhere (albeit with subtle differences).
+    * tokio uses a custom version of `AsyncRead` and `AsyncWrite` traits which differ from the ones used by other parts of the ecosystem.
+* The tokio API is needed to create core runtime operations like timers (`tokio::time::sleep`) and to launch tasks; there doesn't seem to be a standard way to abstract over those kinds of things in a runtime-independent way.
+* Some of their dependencies (e.g `hyper` and `reqwest`) are tied to tokio. In some cases, there are configuration options or ways to use those dependencies that don't depend on tokio, but there is no standard mechanism for that.
+
+These things aren't specific to tokio. There just doesn't seem to be a lot of consensus in the ecosystem on how to write "runtime-independent" code and in some cases there aren't any great options available (e.g., spawning tasks).
+
+They investigate the possibility of providing some sort of compatibility layer between tokio
+and their new runtime of choice but this turns out to not seem like the right way to go as this
+compatibility layer would require too much overhead.
+
+Barbara makes the decision to rewrite large parts of the codebase to use the new runtime. This is
+expensive and takes longer than she originally thought. Barbara is very disappointed.
```

Comment:
Maybe this story doesn't need to end with Barbara rewriting stuff to work with another runtime.  I think the story can still stand if it concludes with Barbara being unsure if making the switch away from tokio is worth it, and being disappointed at the amount of work needed to make the switch.  This would be a bit more consistent with @nikomatsakis's proposal above ("Personally I think it's sufficient to just say that they were investigating what it would take to switch. Maybe she was asked to assess it for her job so she did a good job digging into all the details.")

That said, my comment here is a pretty minor point.

Reply:
I agree, I thnk the story is stronger that way, it represents the real story i've heard most of the time.